tools/astarte_export/lib/export.ex

```diff
@@ -0,0 +1,266 @@
+    %{
+      name: state.interface_name,
+      major_version: state.major_version,
+      minor_version: state.minor_version,
+      active: state.active
+    }
+  end
+
+  defp get_attributes(:datastream, {_type, state}) do
+    %{path: state.path}
+  end
+
+  defp get_attributes(:property, state) do
+    %{path: state.path, reception_timestamp: state.reception_timestamp}
+  end
+
+  defp get_attributes(:object, state) do
+    %{reception_timestamp: state.reception_timestamp}
+  end
+
+  defp get_attributes(:value, state) do
+    %{:reception_timestamp => state[:reception_timestamp]}
+  end
+
+  defp get_attributes(:item, value) do
+    %{name: value.name}
+  end
+
+  defp get_attributes(_tag, _value) do
+    %{}
+  end
+
+  @spec get_value(atom(), struct() | keyword()) ::
+          String.t()
+          | charlist()
+          | maybe_improper_list()
+
+  defp get_value(:device, state) do
+    tag_list = [:protocol, :registration, :credentials, :stats, :interfaces]
+
+    Enum.reduce(tag_list, [], fn tag, acc ->
+      acc ++ [serialize_xml(tag, state)]
+    end)
+  end
+
+  defp get_value(:interfaces, state) do
+    Enum.reduce(state.interfaces, [], fn interface_state, acc ->
+      acc ++ [serialize_xml(:interface, interface_state)]
+    end)
+  end
+
+  defp get_value(:interface, state) do
+    mappings = state.mappings
+    interface_type = state.interface_type
+
+    Enum.reduce(mappings, [], fn mapping, acc ->
+      case interface_type do
+        {:datastream, _} ->
+          acc ++ [serialize_xml(:datastream, mapping)]
+
+        {:properties, _} ->
+          acc ++ [serialize_xml(:property, mapping)]
+      end
+    end)
+  end
+
+  defp get_value(:datastream, mapping) do
+    Enum.reduce(mapping.value, [], fn value, acc ->
+      output =
+        case mapping.aggregation do
+          :object ->
+            serialize_xml(:object, value)
+
+          :individual ->
+            serialize_xml(:value, value)
+        end
+
+      acc ++ [output]
+    end)
+  end
+
+  defp get_value(:object, state) do
+    Enum.reduce(state.value, [], fn value, acc ->
+      [serialize_xml(:item, value) | acc]
+    end)
+  end
+
+  defp get_value(:property, state) do
+    state.value |> Kernel.to_string()
+  end
+
+  defp get_value(:item, state) do
+    state.value |> Kernel.to_string()
+  end
+  
+  defp get_value(:value, state) do
+    state.value |> Kernel.to_string()
+  end
+  
+  defp get_value(tag, _values) do
+    ""
+  end
+
+  def format_time() do
+    {{year, month, date}, {hour, minute, second}} = :calendar.local_time()
+
+    to_string(year) <>
+      "_" <>
+      to_string(month) <>
+      "_" <>
+      to_string(date) <>
+      "_" <>
+      to_string(hour) <>
+      "_" <>
+      to_string(minute) <>
+      "_" <>
+      to_string(second)
+  end
+
+  defp default_xml_header() do
+    XmlBuilder.document("")
+    |> XmlBuilder.generate()
+  end
+
+  defp astarte_default_open_tags do
+    "<astarte>\n<devices>\n"
```

Comment:
What is this?

Reply:
So basically, we are streaming the xml in chunks to the file. XmlBuilder doesn't provide an option to create start_tags alone. Due to which we have to handle the <astrate><devices> tags as default tags and process in first go